### DISCUSSION TITLE
Streaming file

### DISCUSSION THREAD

**Discussion by Paul-Vandell** [salvo-rs/salvo#192] [Q&A] [answered]
Hi there, 
amazing lib gread job here.

Would like to not used the `serve static file` but more having a dynamic file serving. But can't figure out with a simple file serving by streaming an opened file into the response.
I mean just a kind of route `files/<id>/view` where i can just stream the previous saved file from the hard drive.
I saw a `ResBody::Stream()` which require a `futures util stream` but can't figure out from an opened `tokio::io` file into the required stream.
I saw a nice way from `axum` here https://github.com/tokio-rs/axum/discussions/608#discussioncomment-1789020.
Should i try with this example https://github.com/hyperium/hyper/blob/master/examples/send_file.rs but seams to be more in one chunk instead of a stream.

Does this example is the correct way or do we have a better way ?
```
    let mut file = tokio::fs::File::open("image_path.png"))
    .await
    .unwrap();
    let mut contents = vec![];
    file.read_to_end(&mut contents).await.unwrap();
    let t = ResBody::Once(contents.into());
    res.set_body(t);
    res.add_header("Content-Type", "image/png", true).unwrap();
```

What should be your best way here ? Thx in advance.

**Comment by chrislearn** [comment, marked as answer]
Try this:

```rust
#[handler]
async fn hello(res: &mut Response) {
    let file = tokio::fs::File::open("Cargo.toml").await.unwrap();
    let stream = ReaderStream::new(file);

    res.add_header(header::CONTENT_TYPE, "text/toml; charset=utf-8", true);
    res.add_header(header::CONTENT_DISPOSITION,   "attachment; filename=\"Cargo.toml\"", true);
    res.streaming(stream);
}
```
Or you can just use [NamedFiile](https://docs.rs/salvo_core/latest/salvo_core/fs/struct.NamedFile.html).

**Reply by Paul-Vandell** [reply]
Sorry did'nt see the `NamedFiled`, Amazing thx a lot. :)

**Reply by Paul-Vandell** [reply]
@chrislearn  after testing both, the `NamedFile` is 4 time faster than using the `ReaderStream` & `StaticFile` solution. Do you know where the diff come from ?

**Reply by chrislearn** [reply]
I'm not sure, maybe tokio File read file with very small size.
